modules/azure-auth-header/azure-auth-header.c

```diff
@@ -0,0 +1,224 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "azure-auth-header.h"
+#include "driver.h"
+#include "modules/http/http-signals.h"
+#include "messages.h"
+
+#define AZURE_AUTH_HEADER_PLUGIN "azure-auth-header"
+
+struct _AzureAuthHeaderPlugin
+{
+  LogDriverPlugin super;
+
+  gsize   secret_len;
+  guchar *secret;
+  GChecksumType alg;
+
+  gchar   *workspace_id;
+  gchar   *method;
+  gchar   *path;
+  gchar   *content_type;
+};
+
+static gsize
+_get_rfc1123date(gchar *buf, gsize buf_len)
+{
+  time_t now = time(NULL);
+  struct tm gmt;
+
+  gmtime_r(&now, &gmt);
+
+  gchar format[] = "%a, %d %b %Y %H:%M:%S GMT";
+  gsize len = strftime(buf, buf_len, format, &gmt);
+
+  return len;
+}
+
+static GString *
+_azure_auth_header_get_str_to_hash(AzureAuthHeaderPlugin *self, glong content_len, const gchar *date)
+{
+  gchar *fmt = "%s\n%ld\n%s\nx-ms-date:%s\n%s";
+  GString *str = g_string_new(NULL);
+
+  g_string_append_printf(str, fmt, self->method, content_len, self->content_type, date, self->path);
+
+  return str;
+}
+
+static gsize
+_azure_auth_header_get_digest(AzureAuthHeaderPlugin *self, guint8 **digest, const GString *input)
+{
+  gsize digest_len = g_checksum_type_get_length(self->alg);
+  *digest = (guint8 *) g_malloc(sizeof(*digest) * digest_len);
+
+  GHmac *hmac = g_hmac_new(G_CHECKSUM_SHA256, self->secret, self->secret_len);
+  g_hmac_update(hmac, (const guchar *)input->str, -1);
+  g_hmac_get_digest(hmac, *digest, &digest_len);
+
+  g_hmac_unref(hmac);
+  hmac = NULL;
+
+  msg_debug("Azure HTTP signature generated",
+            evt_tag_str("str", input->str),
+            evt_tag_int("len", input->len),
+            evt_tag_int("key_len", self->secret_len));
+
+  return digest_len;
+}
+
+static void
+_azure_auth_header_format_headers(AzureAuthHeaderPlugin *self, List *headers, const gchar *digest, const gchar *date)
+{
+  GString *auth_hdr = g_string_new(NULL);
+  GString *date_hdr = g_string_new(NULL);
+
+  g_string_printf(auth_hdr, "Authorization: SharedKey %s:%s", self->workspace_id, digest);
+  g_string_printf(date_hdr, "x-ms-date: %s", date);
+
+  list_append(headers, auth_hdr->str);
+  list_append(headers, date_hdr->str);
+
+  g_string_free(auth_hdr, TRUE);
+  g_string_free(date_hdr, TRUE);
+}
+
+static gboolean
+_append_headers(AzureAuthHeaderPlugin *self, List *headers, GString *body)
+{
+  g_return_val_if_fail(self->secret, FALSE);
+
+  GString *rawstr = NULL;
+
+  gsize date_max_len = 64;
+  gchar date[date_max_len];
+  gsize date_len = 0;
+
+  gsize   digest_len = 0;
+  guint8 *digest = NULL;
+  gchar  *digest_str = NULL;
+
+  date_len = _get_rfc1123date(date, date_max_len);
```

Comment:
We are not really interested in`date_len`, other than the `stftime` was successful or not inside `_get_rfc1123date`. The assert could be moved there: arount `stftime`.